crates/uv-auth/src/middleware.rs

```diff
@@ -0,0 +1,177 @@
+}
+
+impl AuthMiddleware {
+    pub fn new(keyring_provider: KeyringProvider) -> Self {
+        Self {
+            nrc: Netrc::new().ok(),
+            keyring_provider,
+        }
+    }
+
+    pub fn from_netrc_file(file: &Path, keyring_provider: KeyringProvider) -> Self {
+        Self {
+            nrc: Netrc::from_file(file).ok(),
+            keyring_provider,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Middleware for AuthMiddleware {
+    async fn handle(
+        &self,
+        mut req: Request,
+        _extensions: &mut Extensions,
+        next: Next<'_>,
+    ) -> reqwest_middleware::Result<Response> {
+        let url = req.url().clone();
+        // If the request already has an authorization header, we don't need to do anything.
+        // This gives in-URL credentials precedence over the netrc file.
+        if req.headers().contains_key(reqwest::header::AUTHORIZATION) {
+            if !url.username().is_empty() {
+                AuthenticationStore::save_from_url(&url);
+            }
+            return next.run(req, _extensions).await;
+        }
+
+        // Try auth strategies in order of precedence:
+        if let Some(stored_auth) = AuthenticationStore::get(&url) {
+            // If we've already seen this URL, we can use the stored credentials
+            if let Some(auth) = stored_auth {
+                match auth {
+                    Credential::Basic(_) => {
+                        req.headers_mut().insert(
+                            reqwest::header::AUTHORIZATION,
+                            basic_auth(auth.username(), auth.password()),
+                        );
+                    }
+                    // Url must already have auth if before middleware runs - see `AuthenticationStore::with_url_encoded_auth`
+                    Credential::UrlEncoded(_) => (),
+                }
+            }
+        } else if let Some(auth) = self.nrc.as_ref().and_then(|nrc| {
+            // If we find a matching entry in the netrc file, we can use it
+            url.host_str()
+                .and_then(|host| nrc.hosts.get(host).or_else(|| nrc.hosts.get("default")))
+        }) {
+            let auth = Credential::from(auth);
+            req.headers_mut().insert(
+                reqwest::header::AUTHORIZATION,
+                basic_auth(auth.username(), auth.password()),
+            );
+            AuthenticationStore::set(&url, Some(auth));
+        } else if matches!(self.keyring_provider, KeyringProvider::Subprocess) {
+            // If we have keyring support enabled, we check there as well
+            if let Ok(auth) = get_keyring_auth(&url) {
+                // Keyring auth found - save it and return the request
+                req.headers_mut().insert(
+                    reqwest::header::AUTHORIZATION,
+                    basic_auth(auth.username(), auth.password()),
+                );
+                AuthenticationStore::set(&url, Some(auth));
+            }
```

Comment:
If you do the refactor to separate `None` from the `Error` case we should be able to log failures here.

Reply:
What is the correct way to log errors in `uv` codebase, btw?  `warn!` or `warn_user!` maybe?

Reply:
Hm that's tough.

Ideally we'd attach this as a hint to request failures so when we _error_ we'd be able to suggest why. In the meantime, I think a tracing message is the way to go unless it's very actionable for the user to fix the problem. I believe the difference is a tracing message will only show up with `-v` but `warn_user!` will always display.

Reply:
Done